### PR TITLE
Fix PeriodicHandler error

### PR DESCRIPTION
**Pull request by jparisu**
Fix error in PeriodicHandler so it does not call callback if it has finished.
When reload period is too small, it provokes a segfault because calls DDSRouter when it has been destroyed.

Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: include/ddsrouter/event/EventHandler.hpp [hunk 1]
```diff
@@ -60,6 +61,14 @@ class EventHandler
     EventHandler(
             std::function<void(Args...)> callback);
 
+    /**
+     * @brief Destroy the Event Handler object if it is not being executed
+     *
+     * Stop if \c mutex_ is taken until the event callback has been executed, and after
+     * it destroy the object.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/EventHandler.hpp]: Old file version

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: include/ddsrouter/event/impl/EventHandler.ipp [hunk 2]
```diff
@@ -31,55 +32,93 @@ template <typename ... Args>
...
+
+        // Setting new callback (even when it will not be called)
+        callback_ = DEFAULT_CALLBACK_;
+
+        // Call child methods in case they should do something when handler is disabled
+        callback_unset_nts_();
+    }
+    else
+    {
+        logWarning(DDSROUTER_HANDLER, "Unsetting callback from an EventHandler that had yet no callback set.")
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/event/impl/EventHandler.ipp:93]:
```suggestion
        logWarning(DDSROUTER_HANDLER, "Unsetting callback from an EventHandler that had no callback set.")
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: include/ddsrouter/event/EventHandler.hpp [hunk 3]
```diff
@@ -109,48 +119,83 @@ class EventHandler
...
+     * @note This callback should never be called as the callback is not called
+     * unless \c is_callback_set_ is true. However, it is useful for constructor without callback
+     */
+    static const std::function<void(Args...)> DEFAULT_CALLBACK_;
 };
 
-} /* namespace event */
-} /* namespace ddsrouter */
+}         /* namespace event */
+}         /* namespace ddsrouter */
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/EventHandler.hpp:198]: :+1:  (weird this change did pass the uncrustify)

File: include/ddsrouter/event/EventHandler.hpp [hunk 4]
```diff
@@ -109,48 +119,83 @@ class EventHandler
...
+    /**
+     * @brief Whether the callback of this Handler is set
+     *
+     * This variable also references the enable/disable status of the EventHandler.
+     *
+     * Guard by \c wait_mutex_
+     */
     std::atomic<bool> is_callback_set_;
 
+    //! Mutex to block EventHandler execution while lambda is being set or called
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/EventHandler.hpp:161]: :+1:

File: src/cpp/main.cpp [hunk 5]
```diff
@@ -105,7 +106,9 @@ int main(
                 };
 
         // Creating FileWatcher event handler
-        event::FileWatcherHandler file_watcher_handler(filewatcher_callback, file_path);
+        std::cout << file_path << std::endl;
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/main.cpp:113]: My fault

File: include/ddsrouter/event/impl/EventHandler.ipp [hunk 6]
```diff
@@ -31,55 +32,86 @@ template <typename ... Args>
...
 {
-    is_callback_set_.store(false);
-    callback_ = DEFAULT_CALLBACK_;
-    callback_unset_();
+    std::lock_guard<std::recursive_mutex> lock(event_mutex_);
+
+    if (is_callback_set_)
+    {
+        {
+            // Setting callback
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/impl/EventHandler.ipp]: :+1:

File: include/ddsrouter/event/impl/EventHandler.ipp [hunk 7]
```diff
@@ -31,55 +32,86 @@ template <typename ... Args>
...
+    , threads_waiting_(0)
 {
 }
 
 template <typename ... Args>
 void EventHandler<Args...>::set_callback(
         std::function<void(Args...)> callback) noexcept
 {
+    std::lock_guard<std::recursive_mutex> lock(event_mutex_);
+
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/impl/EventHandler.ipp:54]: :+1:

File: include/ddsrouter/event/impl/EventHandler.ipp [hunk 8]
```diff
@@ -31,55 +32,86 @@ template <typename ... Args>
...
+            {
+                // Exit if number of events is bigger than expected n
+                // or if callback is no longer set
+                return number_of_events_registered_ >= n || !is_callback_set_.load();
+            });
+
+        --threads_waiting_;
+    }
+
+    // Return true if the condition has been fulfilled. It could stop due to a unset callback
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/event/impl/EventHandler.ipp:119]: :+1:

File: include/ddsrouter/event/impl/EventHandler.ipp (hunk 2 above)
Comment (jparisu) [reply include/ddsrouter/event/impl/EventHandler.ipp:93]: :+1:

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**